### PR TITLE
Add Couchbase v2 Destination Connector

### PR DESCRIPTION
**Pull request by lokesh-couchbase**
couchbase.com is a NoSQL Key-Value Store Database. Couchbase would like to integrate with unstructured.io.

This PR Focuses on implementing v2 Destination Connector for Couchbase Database.

### CODE REVIEW

File: test_e2e/dest/couchbase.sh [hunk 1]
```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env bash
```
Comment (rbiseck3) [inline test_e2e/dest/couchbase.sh:1]: This file needs to be updated to be executable (`chmod +x test_e2e/dest/couchbase.sh`)

File: unstructured_ingest/v2/processes/connectors/couchbase.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+if TYPE_CHECKING:
+    from couchbase.cluster import Cluster
+
+CONNECTOR_TYPE = "couchbase"
+SERVER_API_VERSION = "1"
+
+
+@dataclass
+class CouchbaseAccessConfig(AccessConfig):
+    connection_string: str = enhanced_field(sensitive=True)
```
Comment (rbiseck3) [inline unstructured_ingest/v2/processes/connectors/couchbase.py:33]: Don't need to mark all the fields within the AccessConfig class as sensitive since the access config field in the ConnectionConfig is sensitive.

File: unstructured_ingest/v2/processes/connectors/couchbase.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        cluster = Cluster(connection_string, options)
+        cluster.wait_until_ready(timedelta(seconds=5))
+        return cluster
+
+    def run(self, contents: list[UploadContent], **kwargs: Any) -> None:
+        elements_dict = []
+        for content in contents:
+            with open(content.path) as elements_file:
+                elements = json.load(elements_file)
+                # Modify the elements to match the expected couchbase format
```
Comment (rbiseck3) [inline unstructured_ingest/v2/processes/connectors/couchbase.py:125]: This sort of manipulation should live in the stager. The uploader itself shouldn't be responsible for modifying the data by the time it gets it.

File: unstructured_ingest/v2/processes/connectors/couchbase.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+                for element in elements:
+                    new_doc = {
+                        element.pop("element_id", None): {
+                            "embedding": element.pop("embeddings", None),
+                            "text": element.pop("text", None),
+                            "metadata": element.pop("metadata", None),
+                            "type": element.pop("type", None),
+                        }
```
Comment (rbiseck3) [inline unstructured_ingest/v2/processes/connectors/couchbase.py:133]: Since you're creating a new dictionary, no need to use `pop` and modify the original dict. Also we want to make sure the key exists so use the concrete `[]` here rather than `pop` or `get`:
```suggestion
                        element["element_id"]: {
                            "embedding": element.get("embeddings", None),
                            "text": element.get("text", None),
                            "metadata": element.get("metadata", None),
                            "type": element.get("type", None),
                        }
```

File: unstructured_ingest/v2/processes/connectors/couchbase.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        from datetime import timedelta
+
+        from couchbase.auth import PasswordAuthenticator
+        from couchbase.cluster import Cluster
+        from couchbase.options import ClusterOptions
+
+        connection_string = username = password = None
+        access_conf = self.connection_config.access_config
+        try:
+            if access_conf.connection_string is not None:
```
Comment (rbiseck3) [inline unstructured_ingest/v2/processes/connectors/couchbase.py:104]: When would this break and lead to the Exception? Also would make sense to have this as a method on the connection config class.

File: unstructured_ingest/v2/processes/connectors/couchbase.py (hunk 2 above)
Comment (rbiseck3) [inline unstructured_ingest/v2/processes/connectors/couchbase.py:33]: Are these all required? If not, we should mark them all as optional with a default of `None` and provide some checks in a `__post_init__` method, checking that either the connection string was provided or the username/password was.

File: unstructured_ingest/v2/processes/connectors/couchbase.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    connection_config: CouchbaseConnectionConfig
+    cluster: Optional["Cluster"] = field(init=False, default=None)
+    connector_type: str = CONNECTOR_TYPE
+
+    def __post_init__(self):
+        self.cluster = self.connect_to_couchbase()
+
+    @requires_dependencies(["couchbase"], extras="couchbase")
+    def connect_to_couchbase(self) -> "Cluster":
+        from datetime import timedelta
```
Comment (rbiseck3) [inline unstructured_ingest/v2/processes/connectors/couchbase.py:95]: This import can live at the top of the file. Only connector specific ones need to live within a method wrapped by `requires_dependencies` decorator.

File: test_e2e/dest/couchbase.sh [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+source "$SCRIPT_DIR"/cleanup.sh
+
+function cleanup() {
+
+   # Remove docker container
+  echo "Stopping Couchbase Docker container"
+  docker-compose -f "$SCRIPT_DIR"/env_setup/couchbase/common/docker-compose.yaml down --remove-orphans
+
+  # Kill couchbase background process
+  pgrep -f couchbase-dest | xargs kill
```
Comment (rbiseck3) [inline test_e2e/dest/couchbase.sh:35]: Why is this needed if couchbase is being run within docker?

File: test_e2e/python/test-ingest-couchbase-output.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+@click.option("--collection", type=str)
+@click.pass_context
+def cli(
+    ctx,
+    username: str,
+    password: str,
+    connection_string: str,
+    bucket: str,
+    scope: str,
+    collection: str,
```
Comment (rbiseck3) [inline test_e2e/python/test-ingest-couchbase-output.py:40]: Doesn't look like these variables are being used.

File: test_e2e/dest/couchbase.sh (hunk 1 above)
Comment (lokesh-couchbase) [reply test_e2e/dest/couchbase.sh:1]: added

File: unstructured_ingest/v2/processes/connectors/couchbase.py (hunk 2 above)
Comment (lokesh-couchbase) [reply unstructured_ingest/v2/processes/connectors/couchbase.py]: removed sensitive

File: unstructured_ingest/v2/processes/connectors/couchbase.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+
+    def run(
+        self,
+        elements_filepath: Path,
+        file_data: FileData,
+        output_dir: Path,
+        output_filename: str,
+        **kwargs: Any,
+    ) -> Path:
+        with open(elements_filepath) as elements_file:
```
[the remark this answers is not in the document]
Comment (lokesh-couchbase) [reply unstructured_ingest/v2/processes/connectors/couchbase.py:76]: Added the element manipulation logic here

File: unstructured_ingest/v2/processes/connectors/couchbase.py (hunk 3 above)
Comment (lokesh-couchbase) [reply unstructured_ingest/v2/processes/connectors/couchbase.py]: Added the element manipulation logic in stager

File: unstructured_ingest/v2/processes/connectors/couchbase.py (hunk 4 above)
Comment (lokesh-couchbase) [reply unstructured_ingest/v2/processes/connectors/couchbase.py]: Updated

File: unstructured_ingest/v2/processes/connectors/couchbase.py (hunk 5 above)
Comment (lokesh-couchbase) [reply unstructured_ingest/v2/processes/connectors/couchbase.py]: Removed unused exception. Instead added a __post_init__

File: unstructured_ingest/v2/processes/connectors/couchbase.py (hunk 2 above)
Comment (lokesh-couchbase) [reply unstructured_ingest/v2/processes/connectors/couchbase.py]: marked as optional with a checker

File: unstructured_ingest/v2/processes/connectors/couchbase.py (hunk 6 above)
Comment (lokesh-couchbase) [reply unstructured_ingest/v2/processes/connectors/couchbase.py]: updated

File: test_e2e/dest/couchbase.sh [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+fi
+
+# shellcheck disable=SC1091
+source "$SCRIPT_DIR"/cleanup.sh
+
+function cleanup() {
+
+   # Remove docker container
+  echo "Stopping Couchbase Docker container"
+  docker-compose -f "$SCRIPT_DIR"/env_setup/couchbase/common/docker-compose.yaml down --remove-orphans
```
[the remark this answers is not in the document]
Comment (lokesh-couchbase) [reply test_e2e/dest/couchbase.sh]: updated

File: test_e2e/dest/couchbase.sh (hunk 7 above)
Comment (lokesh-couchbase) [reply test_e2e/dest/couchbase.sh]: I actually thought this was some unstructured process. Totally my bad her. Removed this

File: unstructured_ingest/v2/processes/connectors/couchbase.py [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+    connector_type: str = CONNECTOR_TYPE
+
+    def __post_init__(self):
+        # verify if required fields are present
+        required_fields = ["connection_string", "username", "password"]
+        for cb_field in required_fields:
+            if (not getattr(self.access_config, cb_field)) or getattr(
+                self.access_config, cb_field
+            ) == "":
+                raise ValueError(f"Missing field in couchbase access config: {cb_field}")
```
Comment (rbiseck3) [inline unstructured_ingest/v2/processes/connectors/couchbase.py:56]: Move this up as a check on the access configs
```suggestion
```

File: unstructured_ingest/v2/processes/connectors/couchbase.py [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+
+CONNECTOR_TYPE = "couchbase"
+SERVER_API_VERSION = "1"
+
+
+@dataclass
+class CouchbaseAccessConfig(AccessConfig):
+    connection_string: Optional[str] = None
+    username: Optional[str] = None
+    password: Optional[str] = None
```
Comment (rbiseck3) [inline unstructured_ingest/v2/processes/connectors/couchbase.py:35]: I was thinking of something like this. Correct me if I'm wrong in what's required from this. 
```suggestion
    password: Optional[str] = None
    
    def __post_init__(self):
        if not self.connection_string and not (self.username and self.password):
            raise ValueError(f"Either connection string or username and password must be set")
```

Comment (rbiseck3) [reply unstructured_ingest/v2/processes/connectors/couchbase.py:35]: Actually what does the connection string look like? Would this contain the username/password info in it? I see in your code to get the connection, all three are required.

File: test_e2e/dest/couchbase.sh [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+OUTPUT_ROOT=${OUTPUT_ROOT:-$SCRIPT_DIR}
+OUTPUT_DIR=$OUTPUT_ROOT/structured-output/$OUTPUT_FOLDER_NAME
+WORK_DIR=$OUTPUT_ROOT/workdir/$OUTPUT_FOLDER_NAME
+DOWNLOAD_DIR=$OUTPUT_ROOT/download/$OUTPUT_FOLDER_NAME
+max_processes=${MAX_PROCESSES:=$(python3 -c "import os; print(os.cpu_count())")}
+DESTINATION_CB_SCOPE="_default"
+DESTINATION_CB_COLLECTION="_default"
+CI=${CI:-"false"}
+
+source "$SCRIPT_DIR"/env_setup/couchbase/common/constants.env
```
Comment (rbiseck3) [inline test_e2e/dest/couchbase.sh:17]: Since we're using docker to simulate the connection, can we simply set these env vars in the test file directly? Usually we use this approach when the env vars contain secret values we don't want to expose.

File: unstructured_ingest/v2/processes/connectors/couchbase.py (hunk 12 above)
Comment (rbiseck3) [reply unstructured_ingest/v2/processes/connectors/couchbase.py:35]: Looking at this again, the only value that needs to be in the `access_config` is the password. Since neither the username nor connection string have sensitive data, these can live in the connection config.

Comment (rbiseck3) [reply unstructured_ingest/v2/processes/connectors/couchbase.py:35]: Also can a user connect to couchbase without a username or password? This will impact if these are required on the config and we might not need the checks in the `__post_init__` method.

File: unstructured_ingest/v2/processes/connectors/couchbase.py (hunk 11 above)
Comment (lokesh-couchbase) [reply unstructured_ingest/v2/processes/connectors/couchbase.py]: Updated

File: unstructured_ingest/v2/processes/connectors/couchbase.py (hunk 12 above)
Comment (lokesh-couchbase) [reply unstructured_ingest/v2/processes/connectors/couchbase.py]: I have updated to only have password be sensitive.

> Also can a user connect to couchbase without a username or password?

No, username and password are separate from connection string and are required fields to connect to couchbase cluster

File: unstructured_ingest/v2/processes/connectors/couchbase.py [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+
+@dataclass
+class CouchbaseAccessConfig(AccessConfig):
+    connection_string: Optional[str] = None
+    username: Optional[str] = None
+    password: str = enhanced_field(sensitive=True, default=None)
+
+    def __post_init__(self):
+        if not self.connection_string and not (self.username and self.password):
+            raise ValueError("Either connection string or username and password must be set")
```
Comment (rbiseck3) [inline unstructured_ingest/v2/processes/connectors/couchbase.py:39]: Now that I understand the use of these fields, since password is the only sensitive value, it's the only thing that should live in `CouchbaseAccessConfig`, with no `__post_init__` require since it's not marked `Optional`. Let's move `username` and `connection_string` to `CouchbaseConnectionConfig`.
```suggestion
@dataclass
class CouchbaseAccessConfig(AccessConfig):
    password: str
```

File: unstructured_ingest/v2/processes/connectors/couchbase.py [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+
+
+@dataclass
+class CouchbaseConnectionConfig(ConnectionConfig):
+    access_config: CouchbaseAccessConfig = enhanced_field(default_factory=CouchbaseAccessConfig)
+    bucket: Optional[str] = None
+    scope: Optional[str] = None
+    collection: Optional[str] = None
+    batch_size: int = 50
+    connector_type: str = CONNECTOR_TYPE
```
Comment (rbiseck3) [inline unstructured_ingest/v2/processes/connectors/couchbase.py:49]: Also are `bucket`, `scope` or `collection` required? If so, let's not mark then as `Optional` with a default value set. 
```suggestion
@dataclass
class CouchbaseConnectionConfig(ConnectionConfig):
    connection_string: str
    username: str
    access_config: CouchbaseAccessConfig = enhanced_field(default_factory=CouchbaseAccessConfig)
    bucket: Optional[str] = None
    scope: Optional[str] = None
    collection: Optional[str] = None
    batch_size: int = 50
    connector_type: str = CONNECTOR_TYPE
```

File: unstructured_ingest/v2/processes/connectors/couchbase.py (hunk 14 above)
Comment (lokesh-couchbase) [reply unstructured_ingest/v2/processes/connectors/couchbase.py]: Makes sense. So I have made all the fields as required (Because they are required to tell exactly where to store data)

File: unstructured_ingest/v2/processes/connectors/couchbase.py (hunk 15 above)
Comment (lokesh-couchbase) [reply unstructured_ingest/v2/processes/connectors/couchbase.py]: I have added few defaults so that it becomes a bit easier for a local couchbase cluster. However, in most cases, these values are needed to be overriden

File: unstructured_ingest/v2/processes/connectors/couchbase.py [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+        logger.info(
+            f"writing {len(elements)} objects to destination "
+            f"bucket, {self.connection_config.bucket} "
+            f"at {self.connection_config.access_config.connection_string}",
+        )
+        bucket = self.cluster.bucket(self.connection_config.bucket)
+        scope = bucket.scope(self.connection_config.scope)
+        collection = scope.collection(self.connection_config.collection)
+
+        for chunk in batch_generator(elements, self.upload_config.batch_size):
```
Comment (rbiseck3) [inline unstructured_ingest/v2/processes/connectors/couchbase.py:140]: Also to improve performance, this can be fanned out using multiprocessing. Take a look at how that's done [here](https://github.com/Unstructured-IO/unstructured-ingest/blob/main/unstructured_ingest/v2/processes/connectors/pinecone.py#L162-L172).

Comment (lokesh-couchbase) [reply unstructured_ingest/v2/processes/connectors/couchbase.py:140]: I don't know if it will be the best way. Because other rate-limiting on couchbase side might happen a bit earlier, but I will have to check more on that. Maybe let's proceed with approach and multi-processing can be added later-on if required?

**Review by rbiseck3 — APPROVED** [review]

**Comment by rbiseck3** [comment]
Need to add an entry to the CHANGELOG file and bump the version to `0.0.4-dev0` in that file and the `__version__.py` file.

**Comment by lokesh-couchbase** [comment]
Added necessary version updates and changelog

**Comment by lokesh-couchbase** [comment]
Hey @rbiseck3 , there was some issue with how docker was creating couchbase and test case was failing because of it. I have fixed it (by not using interactive mode). Can you please re-run the CI. Thanks
